ui/ozone/platform/wayland/wayland_data_device.cc

```diff
@@ -39,29 +105,128 @@ void WaylandDataDevice::RequestSelectionData(const std::string& mime_type) {
   // Ensure there is not pending operation to be performed by the compositor,
   // otherwise read(..) can block awaiting data to be sent to pipe.
   read_from_fd_closure_ =
-      base::BindOnce(&WaylandDataDevice::ReadClipboardDataFromFD,
+      base::BindOnce(&WaylandDataDevice::ReadClipboardData,
                      base::Unretained(this), std::move(fd), mime_type);
+  RegisterSyncCallback();
+}
+
+void WaylandDataDevice::RequestDragData(
+    const std::string& mime_type,
+    base::OnceCallback<void(const std::string&)> callback) {
+  base::ScopedFD fd = drag_offer_->Receive(mime_type);
```

Comment:
can it happen that drag_offer_ is not available by this time? What's about having
while(!drag_offer_)
  wl_display_roundtrip(...)?

Reply:
I didn't know that Wayland has that API. Good to know. Thanks but on this point, RequestDragData() is triggered by OnDrop event and |drag_offer_| is created by OnEnter event. If |drag_offer_| is not available in RequestDragData(), it means that OnDrop event happens without OnEnter event. I think it's important to get correct event sequence from platform for DnD and we can't make DnD work even if we wait for it here when event sequence is wrong and |drag_offer_| is invalid. WDYT?

Reply:
ok.